File: 20.ValidParenthesis.py
class Solution(object):
    def isValid(self, s):

        stack = []
        for c in s:
            if c in '({[':
                stack.append(c)
            else:
                if not stack or \
                    (c==')' and stack[-1]!='(') or \
                    (c=='}' and stack[-1]!='{') or \
                    (c==']' and stack[-1]!='['):
                    return 0
                stack.pop()
        return not stack

File: .cpp/20.ValidParenthesis.cpp
class Solution
{
public:
    bool isValid(string s)
    {
        if (s.size() % 2 != 0)
        {
            return 0;
        }
        stack<char> st;
        for (char c : s)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                st.push(c);
            }
            else
            {
                if (st.empty() || (c == ')' && st.top() != '(') || (c == ']' && st.top() != '[') || (c == '}' && st.top() != '{'))
                {
                    return 0;
                }
                st.pop();
            }
        }
        return st.empty();
    }
};